pkg/custompluginmonitor/plugin/plugin.go

```diff
@@ -55,41 +55,53 @@ func (p *Plugin) GetResultChan() <-chan cpmtypes.Result {
 func (p *Plugin) Run() {
 	runTicker := time.NewTicker(*p.config.PluginGlobalConfig.InvokeInterval)
 
-	for {
-		select {
-		case <-runTicker.C:
-			glog.Info("Start to run custom plugins")
+	runner := func() {
+		glog.Info("Start to run custom plugins")
 
-			for _, rule := range p.config.Rules {
-				p.syncChan <- struct{}{}
-				p.Add(1)
+		for _, rule := range p.config.Rules {
+			p.syncChan <- struct{}{}
+			p.Add(1)
 
-				go func(rule *cpmtypes.CustomRule) {
-					defer p.Done()
-					defer func() {
-						<-p.syncChan
-					}()
+			go func(rule *cpmtypes.CustomRule) {
+				defer p.Done()
+				defer func() {
+					<-p.syncChan
+				}()
 
-					start := time.Now()
-					exitStatus, message := p.run(*rule)
-					end := time.Now()
+				start := time.Now()
+				exitStatus, message := p.run(*rule)
+				end := time.Now()
 
-					glog.V(3).Infof("Rule: %+v. Start time: %v. End time: %v. Duration: %v", rule, start, end, end.Sub(start))
+				glog.V(3).Infof("Rule: %+v. Start time: %v. End time: %v. Duration: %v", rule, start, end, end.Sub(start))
 
-					result := cpmtypes.Result{
-						Rule:       rule,
-						ExitStatus: exitStatus,
-						Message:    message,
-					}
+				result := cpmtypes.Result{
+					Rule:       rule,
+					ExitStatus: exitStatus,
+					Message:    message,
+				}
+
+				p.resultChan <- result
+
+				glog.Infof("Add check result %+v for rule %+v", result, rule)
+			}(rule)
+		}
 
-					p.resultChan <- result
+		p.Wait()
+		glog.Info("Finish running custom plugins")
+	}
 
-					glog.Infof("Add check result %+v for rule %+v", result, rule)
-				}(rule)
-			}
+	select {
+	case <-p.tomb.Stopping():
+		glog.Info("Stopping plugin execution")
+		p.tomb.Done()
```

Comment:
Perhaps it's better to just move this one `p.tomb.Done()` and the one below to the top, as a `defer p.tomb.Done()`

Reply:
I don't see `Stop()` is invoked anywhere. Do we support graceful shutdown?

But I changed it to use defer.

Reply:
Seems like `tomb.Stop()` is called at here:
https://github.com/kubernetes/node-problem-detector/blame/v0.6.3/pkg/custompluginmonitor/custom_plugin_monitor.go#L84

+cc @andyxning for the context, as he wrote the code.

Reply:
I meant the `customPluginMonitor`'s `Stop()` doesn't seem to be used anywhere

Reply:
Sorry for the late response. Actually `customPluginMonitor`'s `Stop()` is not invoked anywhere. It is just to meet the requirement about implementing `Monitor` interface. 

`Stop` should be called once `p.Run` errors and is about to exit. @yguo0905 Could you take a look at this?